File: app/src/main/java/com/example/olesya/rxjavatest/view/ScreenActivity.java
package com.example.olesya.rxjavatest.view;

import android.app.AlertDialog;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.example.olesya.rxjavatest.Card;
import com.example.olesya.rxjavatest.ClassModels.ServiceHolderActivity;
import com.example.olesya.rxjavatest.R;
import com.example.olesya.rxjavatest.ScreenViewModel;
import com.example.olesya.rxjavatest.Server;
import com.example.olesya.rxjavatest.Utils;
import com.example.olesya.rxjavatest.databinding.ActivityScreenImaginariumBinding;
import com.example.olesya.rxjavatest.interfaces.ScreenCallback;

import java.util.ArrayList;

import jp.wasabeef.recyclerview.animators.SlideInUpAnimator;

public class ScreenActivity extends ServiceHolderActivity implements ScreenCallback {

    private ActivityScreenImaginariumBinding mBinding;
    private ScreenViewModel viewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_screen_imaginarium);
        viewModel = ViewModelProviders.of(this).get(ScreenViewModel.class);
        initListView();
        initCardPager();
        startServerService();
        viewModel.getMessage().observe(this, s -> Utils.showAlert(ScreenActivity.this, s));
        mBinding.buttonSend.setOnClickListener(v -> serviceMessage.postValue("anything"));
//        test();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        stopService(mServiceIntent);
    }

    private void test() {
        viewModel.onSelectedCardEvent(new Card("0", "player"));
        viewModel.onSelectedCardEvent(new Card("1", "player1"));
        viewModel.onSelectedCardEvent(new Card("2", "player2"));
        viewModel.onSelectedCardEvent(new Card("3", "player3"));
    }

    @Override
    public void setCallbacks() {
        ((Server) mService).setCallbacks(viewModel);
        ((Server) mService).setScreenCallbacks(this);
    }

    private void initListView() {
        //init player rv
        mBinding.playersStatusRv.setHasFixedSize(false);
        mBinding.playersStatusRv.setAdapter(viewModel.getPlayerAdapter());
        mBinding.playersStatusRv.setLayoutManager(new LinearLayoutManager(this));
    }

    private void initCardPager() {
        mBinding.cardRv.setHasFixedSize(true);
        mBinding.cardRv.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));
        mBinding.cardRv.setItemAnimator(new SlideInUpAnimator());
        mBinding.cardRv.getItemAnimator().setAddDuration(200);
        mBinding.cardRv.setAdapter(viewModel.getCardAdapter());
        viewModel.getCardAdapter().setItemCallback(this);
        viewModel.getCardAdapter().registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onItemRangeInserted(int positionStart, int itemCount) {
                super.onItemRangeInserted(positionStart, itemCount);
                runOnUiThread(() -> mBinding.cardRv.scrollToPosition(positionStart));
            }

            @Override
            public void onItemRangeRemoved(int positionStart, int itemCount) {
                super.onItemRangeRemoved(positionStart, itemCount);
            }
        });
    }

    protected void startServerService() {
        mServiceIntent = new Intent(this, Server.class);
        if (getIntent().getExtras() != null) {
            int playerNum = getIntent().getExtras().getInt(Utils.CLIENT_NUM);
            int win = getIntent().getExtras().getInt(Utils.WIN_PTS);
            mServiceIntent.putExtra(Utils.CLIENT_NUM, playerNum);
            mServiceIntent.putExtra(Utils.WIN_PTS, win);
        }

        startService(mServiceIntent);
    }

    private void showRoundResults() {
        if (((Server) mService).hasClientWinPts()) {
            Utils.showAlert(this, getResources().getString(R.string.end_of_game));
            ((Server) mService).stopGame();
            return;
        }

        String results = ((Server) mService).getStringResults();
        new AlertDialog.Builder(this, AlertDialog.THEME_DEVICE_DEFAULT_DARK)
                .setTitle(R.string.round_results)
                .setMessage(results)
                .setCancelable(false)
                .setNeutralButton(R.string.OK,
                        (dialog, which) -> onStartNewRound())
                .show();
    }

    @Override
    public void onStartNewRound() {
        ((Server) mService).changeLeader();
        viewModel.startNewRound(mBinding.cardRv);

        new Thread(() -> ((Server) mService).setTurnNextUser()).start();
    }

    @Override
    public void stopRound() {
        ((Server)mService).sendCardsToUsers();
        runOnUiThread(() -> {
            viewModel.showChoices(mBinding.cardRv);
            ((Server) mService).countRoundPts(viewModel.getCardAdapter());
            showRoundResults();
        });
    }

    @Override
    public void onAddUserEvent(String clientName) {
        runOnUiThread(() -> viewModel.getPlayerAdapter().add(clientName));
    }

    @Override
    public void onRemoveUserEvent(String clientName) {
        serviceMessage.postValue(clientName + " out");
        runOnUiThread(() -> viewModel.getPlayerAdapter().removePlayer(clientName));
        ((Server) mService).removePlayer(clientName);
    }

    @Override
    public void initDesk(String username) {
        for (int i = 0; i < Utils.DEFAULT_CARDS_NUM; i++) {
            ((Server)mService).sendRandomCardToUser(username);
        }
    }

    @Override
    public void onShuffleEnd() {
        runOnUiThread(() -> {
            viewModel.unCoverItems(mBinding.cardRv);
            ((Server) getService()).startChoosingStep();
        });
    }
}

File: app/src/main/java/com/example/olesya/boardgames/ui/view/ScreenActivity.kt
package com.example.olesya.boardgames.ui.view

import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import android.content.Intent
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.View
import android.view.animation.Animation
import androidx.recyclerview.widget.RecyclerView
import com.example.olesya.boardgames.Commands
import com.example.olesya.boardgames.R
import com.example.olesya.boardgames.Utils
import com.example.olesya.boardgames.adapter.CardDiffUtil
import com.example.olesya.boardgames.adapter.CardPagerAdapter
import com.example.olesya.boardgames.adapter.PlayerAdapter
import com.example.olesya.boardgames.connection.common.ServiceHolderActivity
import com.example.olesya.boardgames.connection.server.Server
import com.example.olesya.boardgames.databinding.ActivityScreenImaginariumBinding
import com.example.olesya.boardgames.entity.Card
import com.example.olesya.boardgames.ui.viewmodel.ScreenViewModel
import jp.wasabeef.recyclerview.animators.SlideInUpAnimator

class ScreenActivity : ServiceHolderActivity() {

    private lateinit var mBinding: ActivityScreenImaginariumBinding
    private lateinit var viewModel: ScreenViewModel
    private var cardPagerAdapter: CardPagerAdapter = CardPagerAdapter()
    private lateinit var playerStatusAdapter: PlayerAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_screen_imaginarium)
        viewModel = ViewModelProviders.of(this).get(ScreenViewModel::class.java)
        initPlayerStatus()
        initCardPager()

        startServerService()
        initObservers()

    }

    private fun initObservers() {
        mBinding.stars.onStart()
        viewModel.message.observe(this, Observer {
            if (it != null) {
                Utils.showAlert(this, it)
            }
        })

        viewModel.isVisibleStatuses.observe(this, Observer { visible ->
            if (visible != null) {
                mBinding.more.progress = if (visible) 0f else 0.5f
                mBinding.more.setMinProgress(if (visible) 0f else 0.5f)
                mBinding.more.setMaxProgress(if (visible) 0.5f else 1f)
                mBinding.more.playAnimation()
//                mBinding.statusMenu.more.addAnimatorListener(Animation.AnimationListener{ })
            }

        })
    }

    override fun onDestroy() {
        super.onDestroy()
        stopService(mServiceIntent)
    }

    override fun setCallbacks() {
        viewModel.controller = (mService as Server).gameController
        viewModel.controller.test()
        viewModel.controller.lcOwner = this
        viewModel.controller.screenMessage.observe(this, Observer { value ->
            Utils.showSnackbar(mBinding.root, value!!)
        })

        viewModel.controller.screenCards.observe(this) {
            val diffResult = DiffUtil.calculateDiff(CardDiffUtil(cardPagerAdapter.dataset,
                    it.map { it as Card }.toMutableList()))
            cardPagerAdapter.setData(it)
            diffResult.dispatchUpdatesTo(cardPagerAdapter)
            cardPagerAdapter.notifyDataSetChanged()
        }

        viewModel.controller.players.observe(this) {
            playerStatusAdapter.players = it
            playerStatusAdapter.notifyDataSetChanged()
            if (it.size == viewModel.controller.totalPlayers) {
                viewModel.controller.startGame()
            }
        }

        viewModel.controller.screenActions.observe(this, Observer {
            viewModel.handleAction(this, it)
        })
    }

    private fun startServerService() {
        mServiceIntent = Intent(this, Server::class.java)
        if (intent.extras != null) {
            val playerNum = intent.extras.getInt(Commands.CLIENT_NUM)
            val win = intent.extras.getInt(Commands.WIN_PTS)
            mServiceIntent.putExtra(Commands.CLIENT_NUM, playerNum)
            mServiceIntent.putExtra(Commands.WIN_PTS, win)
        }

        startService(mServiceIntent)
    }

    private fun initCardPager() {
        mBinding.cardRv.setHasFixedSize(true)
        mBinding.cardRv.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        mBinding.cardRv.itemAnimator = SlideInUpAnimator()
        (mBinding.cardRv.itemAnimator as SlideInUpAnimator).addDuration = 200
        mBinding.cardRv.adapter = cardPagerAdapter
    }

    private fun initPlayerStatus() {
        playerStatusAdapter = PlayerAdapter()
        mBinding.statusMenu.playersStatusRv.adapter = playerStatusAdapter
        mBinding.more.setOnClickListener({ viewModel.isVisibleStatuses.postValue(!viewModel.isVisibleStatuses.value) })
        mBinding.statusMenu.playersStatusRv.layoutManager =
                LinearLayoutManager(this, RecyclerView.VERTICAL, false)
    }
}